src/main/java/hudson/plugins/pmd/PmdPublisher.java

```diff
@@ -25,14 +24,20 @@
  * @author Ulli Hafner
  */
 public class PmdPublisher extends HealthAwarePublisher {
+
     private static final long serialVersionUID = 6711252664481150129L;
```

Comment:
That's the generic form.

Basically any time you change a class that gets serialized **and** your class introduces new invariants
* e.g. a field which cannot be `null`... there are other potential invariants, but in Jenkins they will all have to be implemented as a field that should not be `null` but is `null` when an older serialized form is deserialized in a newer class.
Well you then need to implement a `readResolve()` method to restore the invariants on deserialization. That `readResolve()` will be called by both XStream and regular java serialization. (you can get more fancy and use other forms of serialization control but XStream can have limited support for some of them).

The most typical use case is where you decide that you want to change the data format... in which case - to handle the transition correctly - you would make the old format fields `transient`. In the `readResolve()` method, if the old form was serialized then the values from the `transient` fields will be populated and thus you check if they were non-`null` and do the translation and populate the new fields (which in this case will be `null` as it was the old serialized form)

The second case is where you want to add a new field that should not be null (now we could argue that a better way is just to have the getter for that field initialize it if it was `null` and always access via the getter... but synchronization costs may mean you might prefer to have the field `/* effectively final */` and thus prefer using the `readResolve()` where you can ignore synchronization as the object instance has not escaped the "constructor" that is deserialization and thus cannot be seen by any other threads - screw you findbugs with your inconsistent synchronization warnings)

HTH